### PR TITLE
Federated ensemble

### PR DESCRIPTION
**Pull request by jewelltaylor**
# PR Type
[**Feature** | Fix | Documentation | Other ]

# Short Description

Clickup Ticket(s): [Federated training ensemble of models](https://app.clickup.com/t/8686ja6he)

Small PR that adds some initial functionality to support training ensembles of models in a federated manner. To this end, I added two classes: 
- `EnsembleModel`: A nn.Module that acts as a wrapper for the ensemble of models with support to produce predictions for individual models as well as an ensemble prediction via voting or averaging.
- `EnsembleClient`: A custom client to handle specific details of training a federated ensemble of models such as setting up the client, computing losses and performing a train step. 

In the process I had to make some minor changes to the `BasicClient` such as store client optimizers in a dictionary  and making the backward attribute of `Losses` a union of Tensor and a dictionary of Tensors. 

It is worth mentioning that I plan on doing a follow up PR that makes it so that we can use Ensembling in conjunction with other client types (ie Scaffold). This seems like a good first start and it would be good to get some eyes on it before I go further. 

# Tests Added
- Add tests to ensure the `EnsembleModel` produces predictions of the current dimension
- Add tests to ensure the "ensemble prediction methods" (ie voting and averaging) work correctly
- A small addition to the tests on `Losses` focused on the new functionality I mentioned

### CODE REVIEW

File: fl4health/clients/apfl_client.py [hunk 1]
```diff
@@ -26,8 +26,7 @@ def __init__(
 
         self.model: ApflModule
         self.learning_rate: float
-        self.optimizer: torch.optim.Optimizer
-        self.local_optimizer: torch.optim.Optimizer
+        self.optimizers: Dict[str, torch.optim.Optimizer]
```
Comment (jewelltaylor) [inline fl4health/clients/apfl_client.py:29]: Update optimizers attribute to be dict instead of a single optimizer which was necessary for the addition of ensemble models.

File: fl4health/clients/basic_client.py [hunk 2]
```diff
@@ -72,7 +72,7 @@ def __init__(
         # Attributes to be initialized in setup_client
         self.parameter_exchanger: ParameterExchanger
         self.model: nn.Module
-        self.optimizer: torch.optim.Optimizer
+        self.optimizers: Dict[str, torch.optim.Optimizer]
```
Comment (jewelltaylor) [inline fl4health/clients/basic_client.py:75]: Update optimizers attribute to be dict instead of a single optimizer which was necessary for the addition of ensemble models.

File: fl4health/clients/basic_client.py [hunk 3]
```diff
@@ -607,7 +612,7 @@ def set_optimizer(self, config: Config) -> None:
         """
         optimizer = self.get_optimizer(config)
         assert not isinstance(optimizer, dict)
-        self.optimizer = optimizer
+        self.optimizers = {"global": optimizer}
```
Comment (jewelltaylor) [inline fl4health/clients/basic_client.py:615]: If we just have a single optimizer, I call it the global one. Maybe this isn't the best name, feel free to suggest something else.

File: fl4health/model_bases/ensemble_base.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+        # If larger than two dimensions, we map to 2D to perfom voting operation (and reshape later)
+        if len(preds_dimension) > 2:
+            preds_list = [preds.reshape(-1, preds_dimension[-1]) for preds in preds_list]
+
+        # For each model prediction, compute the argmax of the model over the classes and stack column-wise into matrix
+        # Each row of matrix represents the argmax of each model for a given sample
+        argmax_per_model = torch.hstack([torch.argmax(preds, dim=1, keepdim=True) for preds in preds_list])
+        # For each row (sample), compute the unique class predictions and their respective counts
+        index_count_list = map(lambda x: torch.unique(x, return_counts=True), argmax_per_model.unbind())
```
Comment (jewelltaylor) [inline fl4health/model_bases/ensemble_base.py:86]: Wasn't sure how to do this entirely in torch because unique returns a tuple. Feel free to suggest something more efficient.

File: fl4health/utils/losses.py [hunk 5]
```diff
@@ -11,7 +11,7 @@ class Losses:
     def __init__(
         self,
         checkpoint: torch.Tensor,
-        backward: torch.Tensor,
+        backward: Union[torch.Tensor, Dict[str, torch.Tensor]],
```
Comment (jewelltaylor) [inline fl4health/utils/losses.py:14]: Backward loss is a dictionary in the case of ensembles.

File: fl4health/utils/losses.py [hunk 6]
```diff
@@ -74,7 +80,20 @@ def compute(self) -> Losses:
         num_losses = len(self.losses_list)
         # Compute average checkpoint and backward losses across list
         checkpoint_loss = torch.sum(torch.FloatTensor([losses.checkpoint for losses in self.losses_list])) / num_losses
-        backward_loss = torch.sum(torch.FloatTensor([losses.backward for losses in self.losses_list])) / num_losses
+        backward_loss: Union[torch.Tensor, Dict[str, torch.Tensor]]
```
Comment (jewelltaylor) [inline fl4health/utils/losses.py:83]: my py was giving me some trouble unless I had these somewhat redundant asserts.

File: examples/ensemble_example/client.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+class MnistEnsembleClient(EnsembleClient):
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        batch_size = self.narrow_config_type(config, "batch_size", int)
+        sampler = DirichletLabelBasedSampler(list(range(10)), sample_percentage=0.75)
+        train_loader, val_loader, _ = load_mnist_data(self.data_path, batch_size, sampler=sampler)
+        return train_loader, val_loader
+
+    def get_model(self, config: Config) -> nn.Module:
+        ensemble_models: Dict[str, nn.Module] = {
```
Comment (emersodb) [inline examples/ensemble_example/client.py:29]: Super unimportant, but do you need the type annotation for this dictionary?

File: fl4health/clients/basic_client.py [hunk 8]
```diff
@@ -531,7 +532,11 @@ def setup_client(self, config: Config) -> None:
         self.num_val_samples = len(self.val_loader.dataset)  # type: ignore
 
         self.set_optimizer(config)
-        self.learning_rate = self.optimizer.defaults["lr"]
+        if "global" in self.optimizers:
```
Comment (emersodb) [inline fl4health/clients/basic_client.py:535]: Should we make `learning_rate` and optional parameter and have clients only set it when it makes sense to use? That is, the only default client that needs it seems to be SCAFFOLD and we can set it by overriding `setup_client` and then doing `self.optimizers["global"].defaults["lr"]`?

File: fl4health/clients/apfl_client.py [hunk 9]
```diff
@@ -101,10 +101,9 @@ def compute_loss(
         return losses
 
     def set_optimizer(self, config: Config) -> None:
-        optimizer_dict = self.get_optimizer(config)
-        assert isinstance(optimizer_dict, dict)
-        self.optimizer = optimizer_dict["global"]
-        self.local_optimizer = optimizer_dict["local"]
+        optimizers = self.get_optimizer(config)
+        assert isinstance(optimizers, dict)
```
Comment (emersodb) [inline fl4health/clients/apfl_client.py:105]: We should probably assert that the dictionary is of length two and contains the proper keys here?

File: fl4health/clients/basic_client.py (hunk 8 above)
Comment (emersodb) [reply fl4health/clients/basic_client.py:535]: This would also avoid the "awkward" setting of the learning rate here to be the "first" optimizer in the dictionary, which I think is a bit "confusing" and sort of implicitly assumes that all optimizer learning rates are equally valid for this value?

File: fl4health/clients/basic_client.py (hunk 3 above)
Comment (emersodb) [reply fl4health/clients/basic_client.py:615]: I don't have a good alternative. Global is good for now.

File: fl4health/clients/basic_client.py (hunk 8 above)
Comment (jewelltaylor) [reply fl4health/clients/basic_client.py:535]: Yeah this is a better approach for sure

File: fl4health/clients/ensemble_client.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        optimizers = self.get_optimizer(config)
+        assert isinstance(optimizers, dict)
+        self.optimizers = optimizers
+
+    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> Tuple[Losses, Dict[str, torch.Tensor]]:
+        """
+        Given a single batch of input and target data, generate predictions
+        (both individual models and ensemble prediction), compute loss, update parameters and
+        optionally update metrics if they exist. (ie backprop on a single batch of data).
+        Assumes self.model is in train model already. Differs from parent method in that, there are multiple losses
```
Comment (emersodb) [inline fl4health/clients/ensemble_client.py:81]: "Assumes self.model is in train model already" to "Assumes self.model is in train mode already"

File: fl4health/utils/losses.py [hunk 11]
```diff
@@ -21,7 +21,13 @@ def __init__(
     def as_dict(self) -> Dict[str, float]:
         loss_dict: Dict[str, float] = {}
         loss_dict["checkpoint"] = float(self.checkpoint.item())
-        loss_dict["backward"] = float(self.backward.item())
+
```
Comment (emersodb) [inline fl4health/utils/losses.py:24]: In the previous iteration, we had to do a float(...) conversion. Is that no longer necessary?

File: fl4health/utils/losses.py (hunk 6 above)
Comment (emersodb) [reply fl4health/utils/losses.py:83]: This is pretty awkward. Maybe, we make the backward value a dictionary as well and have a default key when there is only "one" backward loss. Something like "backward" maybe? That way basic client and any others can just rely on that being there. Then we don't have to do as much here and the types are "unified"?

Comment (emersodb) [reply fl4health/utils/losses.py:83]: If we do this, the averaging procedure would be the same as `additional_losses`. We should encapsulate the averaging functionality in that case into a class function.

File: fl4health/model_bases/ensemble_base.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        aggregation_mode: Optional[EnsembleAggregationMode] = EnsembleAggregationMode.AVERAGE,
+    ) -> None:
+        """
+        Class that acts a wrapper to an ensemble of models to be trained in federated manner with support
+        for both voting and averaging prediction of individual models.
+
+        Args:
+            ensemble_models (Dict[str, nn.Module]): A dictionary of models that make up the ensemble.
+            aggregation_mode (Optional[EnsembleAggregationMode]): The mode in which to aggregate the
+                predictions of indivdual models.
```
Comment (emersodb) [inline fl4health/model_bases/ensemble_base.py:26]: indivdual -> individual 🙂

File: fl4health/utils/losses.py (hunk 5 above)
Comment (lotif) [reply fl4health/utils/losses.py:14]: This could be added as a code comment to explain the `Union`.

File: fl4health/utils/losses.py [hunk 13]
```diff
@@ -74,7 +80,20 @@ def compute(self) -> Losses:
...
+            assert isinstance(self.losses_list[0].backward, dict)
+            backward_loss = {
+                key: torch.sum(
+                    torch.FloatTensor(
+                        [losses.backward[key] for losses in self.losses_list if isinstance(losses.backward, dict)]
+                    )
+                )
+                / num_losses
+                for key in self.losses_list[0].backward.keys()
+            }
```
Comment (lotif) [inline fl4health/utils/losses.py:94]: IMHO this would be more easily readable as a regular for loop than a dictionary comprehension.

File: fl4health/utils/losses.py [hunk 14]
```diff
@@ -74,7 +80,20 @@ def compute(self) -> Losses:
...
         # Compute average checkpoint and backward losses across list
         checkpoint_loss = torch.sum(torch.FloatTensor([losses.checkpoint for losses in self.losses_list])) / num_losses
-        backward_loss = torch.sum(torch.FloatTensor([losses.backward for losses in self.losses_list])) / num_losses
+        backward_loss: Union[torch.Tensor, Dict[str, torch.Tensor]]
+        if all(isinstance(losses.backward, dict) for losses in self.losses_list):
+            assert isinstance(self.losses_list[0].backward, dict)
+            backward_loss = {
+                key: torch.sum(
+                    torch.FloatTensor(
+                        [losses.backward[key] for losses in self.losses_list if isinstance(losses.backward, dict)]
```
Comment (lotif) [inline fl4health/utils/losses.py:89]: What happens if this `if` condition is not satisfied?

**Review by lotif — COMMENTED** [review]
You might as well add this to the smoke tests as well now that it has been merged 😬

File: fl4health/model_bases/ensemble_base.py (hunk 4 above)
Comment (emersodb) [reply fl4health/model_bases/ensemble_base.py:86]: I think this looks pretty good

File: tests/models/test_ensemble_base.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+import torch
+import torch.nn as nn
+
+from fl4health.model_bases.ensemble_base import EnsembleAggregationMode, EnsembleModel
+from tests.test_utils.models_for_test import SmallCnn
+
+
+def test_forward_average_mode() -> None:
+    models: Dict[str, nn.Module] = {"model_0": SmallCnn(), "model_1": SmallCnn()}
+    em = EnsembleModel(models, EnsembleAggregationMode.AVERAGE)
```
Comment (emersodb) [inline tests/models/test_ensemble_base.py:13]: Maybe expand these variable names. em -> ensemble_model and ep -> ensemble_predictions

File: tests/models/test_ensemble_base.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    data = torch.rand((64, 1, 28, 28))
+    ep = em(data)
+
+    assert len(ep) == 3
+    assert "model_0" in ep and ep["model_1"].shape == torch.Size([64, 32])
```
Comment (emersodb) [inline tests/models/test_ensemble_base.py:18]: I think you want this key to be `"model_0"`?

Comment (emersodb) [reply tests/models/test_ensemble_base.py:18]: Similarly below, I think you want to test that `"model_0" in ep`?

File: tests/models/test_ensemble_base.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+def test_forward_vote_mode() -> None:
+    models: Dict[str, nn.Module] = {"model_0": SmallCnn(), "model_1": SmallCnn()}
+    em = EnsembleModel(models, EnsembleAggregationMode.VOTE)
+    data = torch.rand((64, 1, 28, 28))
+    ep = em(data)
+
+    assert len(ep) == 3
+    assert "model_0" in ep and ep["model_0"].shape == torch.Size([64, 32])
+    assert "model_1" in ep and ep["model_0"].shape == torch.Size([64, 32])
```
Comment (emersodb) [inline tests/models/test_ensemble_base.py:31]: key probably should be model_1 here?

File: tests/models/test_ensemble_base.py [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+    assert len(ep) == 3
+    assert "model_0" in ep and ep["model_1"].shape == torch.Size([64, 32])
+    assert "model_0" in ep and ep["model_1"].shape == torch.Size([64, 32])
+    assert "ensemble-pred" in ep and ep["ensemble-pred"].shape == torch.Size([64, 32])
+
+
+def test_forward_vote_mode() -> None:
+    models: Dict[str, nn.Module] = {"model_0": SmallCnn(), "model_1": SmallCnn()}
+    em = EnsembleModel(models, EnsembleAggregationMode.VOTE)
```
Comment (emersodb) [inline tests/models/test_ensemble_base.py:25]: expand these names for readability as well 🙂

File: fl4health/utils/losses.py [hunk 19]
```diff
@@ -58,97 +55,183 @@ class LossMeterType(Enum):
...
+        """
         raise NotImplementedError
 
     @abstractmethod
     def compute(self) -> Losses:
+        """
+        Compute aggregation of current list of losses if non-empty.
+
+        Returns:
+            Losses: New Losses object with average of losses in losses_list.
```
Comment (emersodb) [inline fl4health/utils/losses.py:93]: This doesn't necessarily have to be the "average" of the losses. Could just say aggregation again?

**Review by emersodb — APPROVED** [review]
I think all of my comments look good to go, so I'll approve. I think it's worth addressing one or two of Marcelo's comments that are outstanding and having him review again.

File: examples/ensemble_example/client.py (hunk 7 above)
Comment (jewelltaylor) [reply examples/ensemble_example/client.py:29]: yeah my bad, I meant to mention that its throwing a mypy error if I don't do the type cast with the Dict[str, nn.Module]

File: fl4health/utils/losses.py (hunk 19 above)
Comment (jewelltaylor) [reply fl4health/utils/losses.py:93]: Yup, good catch!

File: examples/ensemble_example/client.py (hunk 7 above)
Comment (emersodb) [reply examples/ensemble_example/client.py:29]: Interesting. Just wanted to check 🙂

File: examples/ensemble_example/client.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+from fl4health.model_bases.ensemble_base import EnsembleModel
+from fl4health.utils.load_data import load_mnist_data
+from fl4health.utils.metrics import Accuracy
+from fl4health.utils.sampler import DirichletLabelBasedSampler
+
+
+class MnistEnsembleClient(EnsembleClient):
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        batch_size = self.narrow_config_type(config, "batch_size", int)
+        sampler = DirichletLabelBasedSampler(list(range(10)), sample_percentage=0.75)
```
[the remark this answers is not in the document]
Comment (jewelltaylor) [reply examples/ensemble_example/client.py]: Added to the config!

File: examples/ensemble_example/client.py [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+        train_loader, val_loader, _ = load_mnist_data(self.data_path, batch_size, sampler=sampler)
+        return train_loader, val_loader
+
+    def get_model(self, config: Config) -> nn.Module:
+        ensemble_models: Dict[str, nn.Module] = {
+            "model_0": ConfigurableMnistNet(out_channel_mult=1).to(self.device),
+            "model_1": ConfigurableMnistNet(out_channel_mult=2).to(self.device),
+            "model_2": ConfigurableMnistNet(out_channel_mult=3).to(self.device),
```
[the remark this answers is not in the document]
Comment (jewelltaylor) [reply examples/ensemble_example/client.py:32]: Not a dumb question at all, its good to have an outside perspective asking these questions :) We can definitely add the number of models as a param in the config. The reason why I chose not too is because people may often want to have non uniform architectures across the models, so I think in most cases they wont be just duplicating the same model. And if its not relevant in all cases, maybe its just best to leave up to the user rather than bake in some convenience functionality that may complicate things. Let me know if you think this is fair or I should add it in!

File: fl4health/utils/losses.py (hunk 5 above)
Comment (jewelltaylor) [reply fl4health/utils/losses.py:14]: Added a comment and better overall documentation!

File: fl4health/clients/scaffold_client.py [hunk 22]
```diff
@@ -45,7 +45,7 @@ def __init__(
         self.client_control_variates: Optional[NDArrays] = None  # c_i in paper
         self.client_control_variates_updates: Optional[NDArrays] = None  # delta_c_i in paper
         self.server_control_variates: Optional[NDArrays] = None  # c in paper
-        self.optimizer: torch.optim.SGD  # Scaffold require vanilla SGD as optimizer
+        self.optimizers: Dict[str, torch.optim.SGD]  # type: ignore
```
[the remark this answers is not in the document]
Comment (jewelltaylor) [reply fl4health/clients/scaffold_client.py:49]: Oops, added back in :)
